Mathlib/Tactic/Linarith/Frontend.lean

```diff
@@ -389,6 +389,13 @@ syntax (name := nlinarith) "nlinarith" "!"? linarithArgsRest : tactic
 @[inherit_doc nlinarith] macro "nlinarith!" rest:linarithArgsRest : tactic =>
   `(tactic| nlinarith ! $rest:linarithArgsRest)
 
+/-- Elaborate `t` in a way that is suitable for linarith. -/
+def elabLinarithArg (tactic : String) (t : Term) : TacticM Expr := Term.withoutErrToSorry do
+  let (e, mvars) ← elabTermWithHoles t none `linarith (allowNaturalHoles := true)
```

Comment:
Hmm, why not use `allowNaturalHoles := false`? That way, the user can see which `_`'s couldn't be synthesized, if that's the issue. (It also might change the behavior just ever so slightly, since `_`'s are elaborated as synthetic opaque holes when `allowNaturalHoles := true`, which iirc changes how delayed metavariables and some other things are constructed.)

Reply:
In the `linarith [p ?a]` test with `?a` as `_`, it wasn't flagging the metavariable for some reason, but allowing natural holes made the check work.

If you have some time, maybe you want to look into this?

We could also just remove the metavariable check. The `Term.withoutErrToSorry` is much more important.

Reply:
In the meantime, I'm going to switch to your suggestion and ignore the issue with `p _`

Reply:
Interesting, I'll take a look right now...

Reply:
So, the issue is the `commitIfNoEx` which rewinds the state if the elaboration fails. We can simply elaborate outside the `commitIfNoEx` if we want to, and that works:
```
elab_rules : tactic
  | `(tactic| linarith $[!%$bang]? $[$cfg]? $[only%$o]? $[[$args,*]]?) => withMainContext do
    let args ← ((args.map (TSepArray.getElems)).getD {}).mapM (elabLinarithArg `linarith)
    let cfg := (← elabLinarithConfig (mkOptionalNode cfg)).updateReducibility bang.isSome
    commitIfNoEx do liftMetaFinishingTactic <| Linarith.linarith o.isSome args.toList cfg
```
I've pushed a commit for convenience in case this seems alright.

It's a bit odd to me that `commitIfNoEx` behaves this way, though. I need to run at the moment, but it might be worth thinking about this more closely...

Reply:
Thanks, I'm happy with this fix -- the elaboration really should go ahead of the "MetaM" part, since that's sort of part of the configuration, and we shouldn't proceed as normal unless that all checks out. I copied these fixes to nlinarith too.

I don't think there's need to think too hard about `commitIfNoEx` itself, but it would be worth investigating why Scott added `commitIfNoEx` here (which part of the state are we wanting to roll back in particular?), and also what other tactics do (if anything) to preserve the message log. Maybe there's already a function for this.

Reply:
I found [this zulip thread](https://leanprover.zulipchat.com/#narrow/stream/348111-std4/topic/Problem.20with.20.60repeat'.60) that led to the change. It seems that it was meant to work around `repeat'` behavior, although Mario fixed `repeat'`, so it wound up being a redundancy measure.

(Also, it looks like the initial absence of `commitIfNoEx` in `nlinarith` may have been an oversight, so I think it makes sense to add it.)

In general, it makes sense to me to roll back state changes made by failed opaque processes which may leave incomprehensible states behind. ("Opaque" here also meaning that it doesn't directly refer to user-entered syntax, as elaboration might, in addition to the usual sense of being unpredictable/a black-box process.) That heuristic by itself is probably not sufficient for a "policy" on when and when not to rewind the state, though, as brought up in the thread.

Also, this is a tangent, but: I'd kind of like to make a document detailing all the different ways we can construct errors in Lean, and the effects that each kind of error has. We can log errors, then continue as usual (unless something else checks if we have errors!); we can totally abort a tactic proof (as we do when there are unassigned natural mvars in `elabTermWithHoles (allowNaturalMvars := false)`); we can `throwError` and `throwErrorAt` (the latter of which might behave differently in subtle ways, iirc?); we can simply produce a `failure` (as with `guard`); we can produce `sorries` when errors occur. And then there's how different processes, like backtracking/try-catch-finally/`observing?`, may or may not rewind (certain) messages or errors in a monad-dependent way, etc.

Anyway, I think this PR looks good now! :)